core/storage/statistics/gae_models.py

```diff
@@ -934,6 +934,117 @@ def save_multi(cls, exploration_stats_dicts):
         cls.put_multi(exploration_stats_models)
 
 
+class ExplorationIssuesModel(base_models.BaseModel):
+    """Model for storing the list of playthroughs for an exploration grouped by
+    issues.
+    """
+    # The unresolved issues for this exploration. This will be a list of dicts
+    # where each dict represents an issue along with the associated
+    # playthroughs.
+    unresolved_issues = ndb.JsonProperty(default=None)
+
+    @classmethod
+    def create(cls, exp_id, unresolved_issues):
+        """Creates an ExplorationIssuesModel instance and writes it to the
+        datastore.
+
+        Args:
+            exp_id: str. ID of the exploration.
+            unresolved_issues: list(dict). The unresolved issues for this
+                exploration. This will be a list of dicts where each dict
+                represents an issue along with the associated playthroughs.
+
+        Returns:
+            str. ID of the new ExplorationIssuesModel instance.
+        """
+        exp_issues_instance = cls(
+            id=exp_id, unresolved_issues=unresolved_issues)
+        exp_issues_instance.put()
+        return exp_id
+
+
+class PlaythroughModel(base_models.BaseModel):
+    """Model for storing recorded useful playthrough data in the datastore.
+
+    The ID of instances of this class are of the form
+    {{exp_id}}.{{random_hash_of_16_chars}}
+    """
+    # ID of the exploration.
+    exp_id = ndb.StringProperty(indexed=True)
+    # Version of the exploration.
+    exp_version = ndb.IntegerProperty(indexed=True)
+    # ID of the issue.
+    issue_id = ndb.StringProperty(indexed=True)
+    # The customization args dict for the given issue_id.
+    issue_customization_args = ndb.JsonProperty(default=None)
+    # The playthrough actions for this playthrough. This will be a list of dicts
+    # where each dict represents a single playthrough action. The list is
+    # ordered by the time of occurence of the action.
+    playthrough_actions = ndb.JsonProperty(default=None)
+    # Boolean for checking validity of the playthrough.
+    is_valid = ndb.BooleanProperty(indexed=True)
+
+    @classmethod
+    def _generate_id(cls, exp_id):
+        """Generates a unique id for the playthrough of the form
+        {{exp_id}}.{{random_hash_of_16_chars}}
+
+        Args:
+            exp_id: str. ID of the exploration.
+
+        Returns:
+            ID of the new PlaythroughModel instance.
+
+        Raises:
+            Exception: The id generator for PlaythroughModel is producing too
+                many collisions.
+        """
+
+        for _ in range(base_models.MAX_RETRIES):
```

Comment:
minor comment: it is a good practice to use xrange over range.

Reply:
Done.